### PR TITLE
Implement text-autospace parser for auto and no-autospace

### PR DESCRIPTION
**Pull request by vitorroriz**
#### a9febb26c871be962432254ecb980a9a36ac7f13
<pre>
Implement text-autospace parser for auto and no-autospace
<a href="https://bugs.webkit.org/show_bug.cgi?id=252119">https://bugs.webkit.org/show_bug.cgi?id=252119</a>
rdar://105340814

Reviewed by Myles C. Maxfield.

text-spacing longhand: text-autospace parser for auto and no-autospace

Reference: <a href="https://github.com/w3c/csswg-drafts/issues/4246#issuecomment-1404738513">https://github.com/w3c/csswg-drafts/issues/4246#issuecomment-1404738513</a>

Implement a parser for text-autospace (text-spacing longhand) for none
and auto-space values. We can then iterate from here add remaining values.
Syntax:
text-autospace: auto | no-autospace

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-list.js:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::textAutospaceFromStyle):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeTextAutospace):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/platform/text/TextSpacing.h:
(WebCore::TextAutospace::isAuto const):
(WebCore::TextAutospace::isNoAutospace const):
(WebCore::TextAutospace::operator== const):
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::textAutospace const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::setTextAutospace):
(WebCore::RenderStyle::initialTextAutospace):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextAutospace):

Canonical link: <a href="https://commits.webkit.org/260306@main">https://commits.webkit.org/260306@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/a7fb3ae488656e79884fd753a1866a000a3fea93

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107902 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16957 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40785 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117035 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116384 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111789 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18346 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8274 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100087 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113664 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97067 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41666 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95762 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28704 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9882 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30052 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10589 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6944 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16034 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49641 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7130 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12161 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebCore/css/CSSProperties.json [hunk 1]
```diff
@@ -1146,6 +1146,25 @@
...
+            "inherited": true,
+            "values": [
+                "auto",
+                "no-autospace"
+            ],
+            "codegen-properties": {
+                "settings-flag": "cssTextSpacingEnabled",
+                "converter": "TextAutospace",
+                "parser-function": "consumeTextAutospace",
+                "parser-grammar-unused": "",
```
Comment (nt1m) [inline Source/WebCore/css/CSSProperties.json]: Can you fill in this field? it's helpful to audit missing features of the new parser

File: Source/WebCore/css/CSSValueKeywords.in [hunk 2]
```diff
@@ -1655,3 +1655,6 @@ only
 // inherits
 true
 false
+
+// for text-autospace
```
Comment (nt1m) [inline Source/WebCore/css/CSSValueKeywords.in]:
```suggestion
// text-autospace
```

File: Source/WebCore/platform/text/TextSpacing.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+#pragma once
+
+namespace WebCore {
+
+struct TextAutospace {
+enum class TextAutospaceType: uint8_t {
+    Auto = 0,
+    NoAutospace // equivalent to None in trim-spacing shorthand
```
Comment (nt1m) [inline Source/WebCore/platform/text/TextSpacing.h]: I don't see any property named trim-spacing in the spec.

File: Source/WebCore/rendering/style/RenderStyle.cpp [hunk 4]
```diff
@@ -2673,9 +2678,9 @@ void RenderStyle::setCustomPropertyValue(Ref<const CSSCustomPropertyValue>&& val
     auto set = [&](auto& data) {
         auto& name = value->name();
         auto* existingValue = data->customProperties->values.get(name);
-        if (existingValue && existingValue->equals(value.get()))
-            return;
-        data.access().customProperties.access().setCustomPropertyValue(name, WTFMove(value));
+    if (existingValue && existingValue->equals(value.get()))
+        return;
+    data.access().customProperties.access().setCustomPropertyValue(name, WTFMove(value));
```
Comment (nt1m) [inline Source/WebCore/rendering/style/RenderStyle.cpp]: unrelated change

**Review by nt1m — COMMENTED** [review]
This new property needs animation support (discrete animation probably) in CSSPropertyAnimation.cpp .

File: Source/WebCore/animation/CSSPropertyAnimation.cpp [hunk 5]
```diff
@@ -3492,6 +3493,7 @@ CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap()
         new DiscretePropertyWrapper<WhiteSpace>(CSSPropertyWhiteSpace, &RenderStyle::whiteSpace, &RenderStyle::setWhiteSpace),
         new DiscretePropertyWrapper<WordBreak>(CSSPropertyWordBreak, &RenderStyle::wordBreak, &RenderStyle::setWordBreak),
         new DiscretePropertyWrapper<OverflowAnchor>(CSSPropertyOverflowAnchor, &RenderStyle::overflowAnchor, &RenderStyle::setOverflowAnchor),
+        new DiscretePropertyWrapper<TextAutospace>(CSSPropertyTextAutospace, &RenderStyle::textAutospace, &RenderStyle::setTextAutospace),
```
Comment (graouts) [inline Source/WebCore/animation/CSSPropertyAnimation.cpp:3497]: It doesn't look like there is testing coverage for this. Could you add the appropriate entry to [web-animations/animation-model/animation-types/property-list.js](https://github.com/web-platform-tests/wpt/blob/master/web-animations/animation-model/animation-types/property-list.js)? Should be something like this:

```javascript
  'align-content': {
    // https://drafts.csswg.org/link-to-spec
    types: [
      { type: 'discrete' , options: [ [ 'normal', 'auto' ] ] }
    ]
  },
```

You can feel free to expand on that of course.

Comment (vitorroriz) [reply Source/WebCore/animation/CSSPropertyAnimation.cpp:3497]: added, thanks

File: Source/WebCore/rendering/style/RenderStyle.cpp (hunk 4 above)
Comment (vitorroriz) [reply Source/WebCore/rendering/style/RenderStyle.cpp:2683]: Style checker was complaining, it seems someone forgot to unintend the line after the return.

**Comment by graouts** [comment]
You need to export those changes to the WPT repository. You can automate that using `export-w3c-test-changes`.

**Comment by vitorroriz** [comment]
> You need to export those changes to the WPT repository. You can automate that using `export-w3c-test-changes`.

We don't want to do that yet. The spec is in flux and we are implementing partially based on the CSSWG resolution not yet converted into the draft.

**Review by litherum — APPROVED** [review]
